### PR TITLE
Fixes for kernel 6.3+

### PR DESCRIPTION
**Pull request by Thesola10**

### CODE REVIEW

**Comment by Thesola10** [comment]
i think we may want to finally deprecate ashmem. Upstream doesn't have it, and Waydroid doesn't need it anymore with `sys.use_memfd=true` in `waydroid_base.prop`

**Comment by ColaMint** [comment]
If you are using version 6.3+ and only need the binder, you can temporarily install it using the following steps:

Clone the anbox-modules repository by running:
```
git clone https://github.com/choff/anbox-modules.git
```
Change directory to the cloned repository:
```
cd anbox-modules
```
Fetch the local branch of the 10th pull request:
```
git fetch origin pull/10/head:local
```
Check out the local branch:
```
git checkout local
```
Modify the contents of anbox.conf to:
```
binder_linux
```
Modify the contents of 99-anbox.rules to:
```
KERNEL=="binder*", NAME="%k", MODE="0666"
```
Modify the contents of [INSTALL.sh](http://install.sh/) to:
```
#!/usr/bin/env bash

sudo cp anbox.conf /etc/modules-load.d/
sudo cp 99-anbox.rules /lib/udev/rules.d/

sudo cp -rT binder /usr/src/anbox-binder-1

sudo dkms install anbox-binder/1

sudo modprobe binder_linux
lsmod | grep -e binder_linux
ls -alh /dev/binder
```
Run [INSTALL.sh](http://install.sh/).

Restart your system.

**Comment by Lufthoheit** [comment]
Worked perfect for me on pop Os 22.04 with modified linux-surface kernel 6.3.3.

File: binder/binder.c [hunk 1]
```diff
@@ -5234,8 +5234,12 @@ static int binder_mmap(struct file *filp, struct vm_area_struct *vma)
 		       proc->pid, vma->vm_start, vma->vm_end, "bad vm_flags", -EPERM);
 		return -EPERM;
 	}
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,3,0))
+    vm_flags_mod(vma, VM_DONTCOPY | VM_MIXEDMAP, VM_MAYWRITE);
```
Comment (choff) [inline binder/binder.c]: Please fix the indentation here and elsewhere in the patch; the convention in the Linux kernel is to indent with spaces and not tabs.

**Review by choff — COMMENTED** [review]
Thank you very much for your patch! And I am sorry that it took so long to review it. It looks good and the only problem are the indentation problems (tabs vs. spaces) that I mentioned in another comment. If you fix that one issue, I will be happy to merge the code.

**Comment by choff** [comment]
> i think we may want to finally deprecate ashmem. Upstream doesn't have it, and Waydroid doesn't need it anymore with `sys.use_memfd=true` in `waydroid_base.prop`

I'd agree, at some point, we should try to get rid of it. Maybe we should give it one more year or so though to avoid locking Anbox users out (not sure if there are still any, though?).

**Comment by Thesola10** [comment]
Okay, this is odd, a lot of tabs occur on areas I hadn't changed, but still, I replaced them all with 4 spaces.

As for Ashmem, we could merge this and tackle backporting Ashmem separately?

**Comment by Thesola10** [comment]
Also, but it's a small nitpick, the way you merge commits seems to invalidate my commit signatures, could you try a different merge method this time around?

**Comment by choff** [comment]
Please you have now changed the indentation from tabs to spaces. But Linux kernel policy is to use tabs, not spaces (did I say it the wrong way round..?). So, we need tabs, not spaces.

Not sure why your commit signatures were lost, but I'll try my best to prevent that it happens again.

**Review by choff — APPROVED** [review]
Thank you, looks good now :-)

And sorry for the confusion about tabs vs. spaces if it came from me.. ;-)

**Comment by choff** [comment]
> As for Ashmem, we could merge this and tackle backporting Ashmem separately?

Thanks again and I agree, we can tackle Ashmem separately..

**Comment by Thesola10** [comment]
woops, sig's gone again ¯\\\_(ツ)_/¯
image

**Comment by choff** [comment]
Sorry :-(

I didn't know that it came from "rebase and merge" - probably this could be avoided by doing a simple merge without rebase. I generally prefer "rebase and merge" though because that leaves the commit history clean (no merge commit). But I can also make an exception and will try another method next time - please just maybe give me a hint about this again when you set another PR.

**Comment by Thesola10** [comment]
That's reasonable. Also, when it comes to sweeping changes, rebase+merge makes commit hashes unreliable, so it's harder for contributors to mark out commits from blame through `.git-blame-ignore-revs`.

**Pull request merged**